File: segundo/1er cuatri/PDOO/irrgarten_ruby/Irrgarten/labyrinth_character.rb
# frozen_string_literal: true

class LabyrinthCharacter
  def initialize(name, intelligence, strength, health)
    @name = name
    @intelligence = intelligence
    @strength = strength
    @health = health
    @row = 0
    @col = 0
  end

  def self.copy(other)
    new(other.name, other.intelligence, other.strength, other.health, other.row, other.col)
  end

  def dead
    true if @health == 0
  end

  def to_s
    "[#{@name}, inteligencia: #{@intelligence}, fuerza: #{@strength}, vida: #{@health}, fila: #{@row}, columna: #{@col}"
  end

  def got_wounded
    @health -= 1
  end

  attr_accessor :health
  attr_reader :intelligence, :strength, :row, :col
  protected :got_wounded

  def set_pos(row, col)
    @row = row
    @col = col
  end

  def attack
    raise "Metodo abstracto"
  end

  def defend(attack)
    raise "Metodo abstracto"
  end
end

File: segundo/1er cuatri/PDOO/Irrgarten_java/src/irrgarten/LabyrinthCharacter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author alvarogz
 */
public abstract class LabyrinthCharacter {
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;

    public LabyrinthCharacter(String name, float intelligence, float strength, float health) {
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = health;
        this.row = 0;
        this.col = 0;
    }

    public LabyrinthCharacter(LabyrinthCharacter other) {
        this.name = other.name;
        this.intelligence = other.intelligence;
        this.strength = other.strength;
        this.health = other.health;
        this.row = other.row;
        this.col = other.col;
    }

    public int getRow() {
        return this.row;
    }   
    
    public int getCol() {
        return this.col;
    }
    
    public boolean dead() {
        return this.health == 0;
    }

    protected float getIntelligence() {
        return intelligence;
    }

    protected float getStrength() {
        return strength;
    }

    protected float getHealth() {
        return health;
    }

    protected void setHealth(float health) {
        this.health = health;
    }
    
    public void setPos(int row, int col) {
        this.row = row;
        this.col = col;
    }
    
    @Override
    public String toString(){
        return this.name + ", inteligencia " + this.intelligence + ", fuerza: " + this.strength + ", vida: " + this.health + ", fila: " + this.row + ", columna: " + this.col;
    }
    
    protected void gotWounded(){
        this.health--;  
    }
    
    public abstract float attack();
    public abstract boolean defend(float attack);
}
